src/felzenszwalb.jl

```diff
@@ -0,0 +1,184 @@
+"""
+```
+index_map, num_segments = felzenszwalb_graph(edges, num_vertices, k)
+```
+
+Efficient Graph-Based Image Segmentation.
+"""
+function felzenszwalb_graph(edges::Array{SVector{3, Float64}}, num_vertices::Int, k::Real)
+
+    num_edges = length(edges)
+    num_sets = num_vertices
+    rank = zeros(num_sets)
+    set_size = ones(num_sets)
+    parent = Array(range(1, num_sets))
+    threshold = fill(k, num_vertices)
+    
+    function find_root(x)
+        x = Int(x)
+        if parent[x]!=x
+            parent[x] = find_root(parent[x])
+        end
+        return parent[x]
+    end
+
+    function merge_trees(x, y)
+        x = Int(x)
+        y = Int(y)
+        num_sets -= 1
+        if rank[x]<rank[y]
+            parent[x] = y
+            set_size[y] += set_size[x]
+            return y
+        elseif rank[x]>rank[y]
+            parent[y] = x
+            set_size[x] += set_size[y]
+            return x
+        else
+            parent[y] = x
+            set_size[x] += set_size[y]
+            rank[x] += 1
+            return x
+        end
+    end
+
+    sort(edges, lt = (x,y)->(x[3]<y[3]))
+
+    for i in range(1,num_edges)
+        edge = edges[i]
+        a, b, w = edge
+        a = find_root(a)
+        b = find_root(b)
+        if a!=b
+            if w <= min(threshold[a], threshold[b])
+                merged_root = merge_trees(a, b)
+                threshold[merged_root] = w + k/set_size[merged_root]
+            end
+        end
+    end
+
+    segments = OrderedSet()
+    for i in 1:num_vertices
+        push!(segments, find_root(i))
+    end
+
+    segments2index = Dict{Int, Int}()
+    for i in 1:num_sets
+        segments2index[segments[i]]=i 
+    end
+
+    index_map = Array{Int64}(num_vertices)
+    for i in 1:num_vertices
+        index_map[i] = segments2index[find_root(i)]
+    end
+
+    return index_map, num_sets
+end
+
+function felzenszwalb{T<:Gray}(img::AbstractArray{T, 2}, k::Real; sigma=0.8)
+
+    #img = imfilter(img, Kernel.gaussian(sigma))
+    rows, cols = size(img)
+    num_vertices = rows*cols
+    num_edges = 4*rows*cols - 3*rows - 3*cols + 2
+    edges = Array{SVector{3, Float64}}(num_edges)
+
+    R = CartesianRange(size(img))
+    I1, Iend = first(R), last(R)
+    num = 1
+    for I in R
+        for J in CartesianRange(max(I1, I-I1), min(Iend, I+I1))
+            if I >= J
+                continue
+            end
+            edges[num] = SVector((I[2]-1)*rows+I[1], (J[2]-1)*rows+J[1], abs(img[I]-img[J]))
```

Comment:
Seems like the first two mean something different than the second one. How about defining
```julia
struct ImageEdge
    index1::Int
    index2::Int
    weight::Float64
end
```
and creating a `Vector{ImageEdge}` to store your graph?

If you use `sub2ind` I think you can support ND images for "free".